### PR TITLE
Log Rows: Added popover menu with filter options when a log line is selected

### PR DESCRIPTION
**Pull request by matyax**
Easier search and filter through the use of a popover menu. 

It makes use of the preexisting Explore methods to add filters from Log Details, adding this function to the situation where the user highlights part of a log line.

Hghlights:

- Supports:
  - Copy text
  - Add line contains filter
  - Add line does not contain filter
- It is dismissed when:
  -  An option is selected 
  - The user clears the current selection
  - By pressing escape
- Shows up when the user selected text in a log line; doesn't show up when multiple log lines are selected.
- Hides the log row menu while selecting text in a log line.

**Which issue(s) does this PR fix?**:

Part of https://github.com/grafana/grafana/issues/76129

**Related PRs**:

- Filtering functions: https://github.com/grafana/grafana/pull/76757
- Feature flag and tracking: https://github.com/grafana/grafana/pull/77616

**Demos**

Elasticsearch:

https://github.com/grafana/grafana/assets/1069378/edd94e72-d23d-40e9-b6ce-be4253045838

Loki:

https://github.com/grafana/grafana/assets/1069378/2ebb781a-dcf9-411a-8524-8d8da42e4025

### CODE REVIEW

**Comment by matyax** [comment]
Demo updated.

**Comment by gtk-grafana** [comment]
This won't work in safari due to https://github.com/grafana/grafana/issues/74135

**Comment by matyax** [comment]
@grafana/observability-logs This is ready for review ✨

File: public/app/features/logs/components/PopoverMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    <div className={styles.menu} style={{ top: y, left: x }}>
+      <Menu ref={containerRef}>
+        <Menu.Item
+          label="Copy selection"
+          onClick={() => {
+            copyText(selection, containerRef);
+            close();
+          }}
+        />
+        {onClickFilterValue && (
```
Comment (ifrost) [inline public/app/features/logs/components/PopoverMenu.tsx:60]: I think we may need a better API to check what types of filters the given data source supports. At the moment we have a single check in https://github.com/grafana/grafana/blob/main/public/app/features/explore/Logs/LogsContainer.tsx#L91 but a datasource may support only label/tag filter (`ADD_FILTER / ADD_FILTER_OUT`) but not line filters (`ADD_LINE_FILTER / ADD_LINE_FILTER_OUT`) which may lead to showing no-op items in the menu.

We did a very similar thing with Supplementary Queries. A data source implements `getSupportedSupplementaryQueryTypes()` providing list of supported types.

**Comment by matyax** [comment]
@ifrost thanks for the feedback!

> Have you considered extending popover menu with Grafana UI Extensions?

I've no reference of Grafana UI Extensions. What is it?

**Comment by matyax** [comment]
Gotcha! It's definitely a great idea to "explore" 😉

**Comment by svennergr** [comment]
Without doing a code review yet, I noticed that the menu would remain open once the user starts scrolling. I am wondering if we should change that and just hide the menu in that case?

https://github.com/grafana/grafana/assets/8092184/9b837f23-7ffd-4de5-b664-bda477d3da38

File: public/app/features/logs/components/LogRow.tsx [hunk 2]
```diff
@@ -121,6 +127,13 @@ class UnThemedLogRow extends PureComponent<Props, State> {
     }
   };
 
+  onMouseMove = (e: MouseEvent) => {
+    // The user is selecting text, so hide the log row menu so it doesn't interfere.
+    if (document.getSelection()?.toString() && e.buttons > 0) {
+      this.setState({ mouseIsOver: false });
+    }
```
Comment (svennergr) [inline public/app/features/logs/components/LogRow.tsx:134]: Nice!

File: public/app/features/logs/components/LogRows.tsx [hunk 3]
```diff
@@ -59,10 +60,15 @@ export interface Props extends Themeable2 {
...
   overflowingContent?: boolean;
+  onClickFilterValue?: (value: string, refId?: string) => void;
+  onClickFilterOutValue?: (value: string, refId?: string) => void;
 }
 
 interface State {
   renderAll: boolean;
+  selection: string;
+  selectedRow: LogRowModel | null;
+  menuCoordinates: { x: number; y: number };
```
Comment (svennergr) [inline public/app/features/logs/components/LogRows.tsx:71]: Just a nit, to not confuse it with the LogRowMenu:

```suggestion
  popoverMenuCoordinates: { x: number; y: number };
```

File: public/app/features/logs/components/LogRows.tsx [hunk 4]
```diff
@@ -85,6 +94,42 @@ class UnThemedLogRows extends PureComponent<Props, State> {
...
+  };
+
+  handleDeselection = (e: Event) => {
+    if (targetIsElement(e.target) && e.target?.getAttribute('role') === 'menuitem') {
+      // Delegate closing the menu to the popover component.
+      return;
+    }
+    if (document.getSelection()?.toString()) {
+      return;
+    }
```
Comment (svennergr) [inline public/app/features/logs/components/LogRows.tsx:121]: I was wondering if we could somehow scope that to the log rows? Because now, basically every selection will keep the menu open:


https://github.com/grafana/grafana/assets/8092184/6dc79f84-4d0c-43e1-bb96-8af57c05b7ae

File: public/app/features/logs/components/LogRow.tsx (hunk 2 above)
Comment (svennergr) [reply public/app/features/logs/components/LogRow.tsx:134]: Just noticed in dashboards, that the copy button isn't reappearing, which makes sense, but just want to let that here. Maybe something we can debounce after a certain timeout:


https://github.com/grafana/grafana/assets/8092184/6bb1302f-7a5f-4a0f-87af-4c8ddb6ac47f

File: public/app/features/logs/utils.ts [hunk 5]
```diff
@@ -273,3 +273,27 @@ export const getLogsVolumeDataSourceInfo = (dataFrames: DataFrame[]): { name: st
 export const isLogsVolumeLimited = (dataFrames: DataFrame[]) => {
   return dataFrames[0]?.meta?.custom?.logsVolumeType === LogsVolumeType.Limited;
 };
+
+export const copyText = async (text: string, buttonRef: React.MutableRefObject<Element | null>) => {
+  if (navigator.clipboard && window.isSecureContext) {
```
Comment (svennergr) [inline public/app/features/logs/utils.ts:278]: I was wondering if we need the fallback? 
`writeText` seems to be supported by every browser since latest early 2020: https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/writeText

File: public/app/plugins/datasource/loki/modifyQuery.ts [hunk 6]
```diff
@@ -540,14 +540,14 @@ function addLabelFormat(
...
-export function addLineFilter(query: string): string {
+export function addLineFilter(query: string, value = ''): string {
   const streamSelectorPositions = getStreamSelectorPositions(query);
   if (!streamSelectorPositions.length) {
     return query;
   }
   const streamSelectorEnd = streamSelectorPositions[0].to;
 
-  const newQueryExpr = query.slice(0, streamSelectorEnd) + ' |= ``' + query.slice(streamSelectorEnd);
+  const newQueryExpr = query.slice(0, streamSelectorEnd) + ` |= \`${value}\`` + query.slice(streamSelectorEnd);
```
Comment (svennergr) [inline public/app/plugins/datasource/loki/modifyQuery.ts:550]: I think this needs some escaping of the value:

https://github.com/grafana/grafana/assets/8092184/84fe22a6-e6ec-4cb5-8c84-f57684fc4c9c

**Review by svennergr — COMMENTED** [review]
Nice feature, really like it. Left a few comments.

**Comment by matyax** [comment]
> Without doing a code review yet, I noticed that the menu would remain open once the user starts scrolling.

Good point! We can definitely hide it on those situations.

File: public/app/features/logs/components/LogRow.tsx (hunk 2 above)
Comment (matyax) [reply public/app/features/logs/components/LogRow.tsx:134]: Ohh, good catch!

Comment (matyax) [reply public/app/features/logs/components/LogRow.tsx:138]: This has been fixed.

File: public/app/features/logs/components/LogRows.tsx (hunk 3 above)
Comment (matyax) [reply public/app/features/logs/components/LogRows.tsx:71]: Done.

**Comment by matyax** [comment]
@svennergr To close the menu while scrolling I would need to listen to the particular scroll event of the `scrollElement` used by Explore, and I'm not sure it's worth the effort of passing down the prop so it reaches the component. Instead, I'm allowing the menu to scroll along with the log lines.

https://github.com/grafana/grafana/assets/1069378/beb56d47-fb16-4e9a-ac63-3860af7f65d8

What do you think?

File: public/app/features/logs/components/LogRows.tsx (hunk 4 above)
Comment (matyax) [reply public/app/features/logs/components/LogRows.tsx:130]: Now closing then the selection event comes from outside of the log rows:

https://github.com/grafana/grafana/assets/1069378/eb545e93-1323-4595-8672-6ae93adbc3f8

File: public/app/features/logs/components/PopoverMenu.tsx (hunk 1 above)
Comment (matyax) [reply public/app/features/logs/components/PopoverMenu.tsx:60]: I think at some point we should deprecate checking for `modifyQuery` in favor of `hasToggleableQueryFiltersSupport()` which is the desired behavior (and API) for log details filters.

File: public/app/plugins/datasource/loki/modifyQuery.ts (hunk 6 above)
Comment (matyax) [reply public/app/plugins/datasource/loki/modifyQuery.ts:550]: Good catch! Will do in the other PR.

File: public/app/features/logs/components/getLogRowStyles.ts [hunk 7]
```diff
@@ -67,12 +67,16 @@ export const getLogRowStyles = memoizeOne((theme: GrafanaTheme2) => {
       color: ${theme.components.textHighlight.text}
       background-color: ${theme.components.textHighlight};
     `,
```
Comment (matyax) [inline public/app/features/logs/components/getLogRowStyles.ts:69]: I should ideally migrate this whole module to use object-type style declarations, but I prefer to do it in a separate PR.

**Comment by matyax** [comment]
@svennergr FYI I merged all the sub-PRs into this one.

File: public/app/features/logs/components/PopoverMenu.tsx (hunk 1 above)
Comment (ifrost) [reply public/app/features/logs/components/PopoverMenu.tsx:60]: Are you making more changing around this in this PR? I think at the moment the popover will show for data sources that may not support `ADD_LINE_FILTER` but they support `ADD_FILTER` 🤔

Comment (matyax) [reply public/app/features/logs/components/PopoverMenu.tsx:60]: I will follow up this PR with a fix for this. Something similar to `getSupportedSupplementaryQueryTypes()` would make an interesting approach, as I don't think we need a new interface for this.

**Review by svennergr — APPROVED** [review]
LGTM, just one nit: sometimes the menu does not disappear after left-clicks on the selected text.

https://github.com/grafana/grafana/assets/8092184/8e9b1d53-8dea-4fa8-9e82-41edacd81527

**Comment by matyax** [comment]
Had to cheery-pick the commits from the feature flag and tracking because they were missing in the other PR, probably due to a bad rebase.

**Comment by matyax** [comment]
Had to cheery-pick the commits from the feature flag and tracking because they were missing in the other PR, probably due to a bad rebase.

Will follow up with a PR to address:
- https://github.com/grafana/grafana/pull/75306#discussion_r1383279017
- https://github.com/grafana/grafana/pull/75306#pullrequestreview-1732218133

File: public/app/features/logs/components/PopoverMenu.tsx (hunk 1 above)
Comment (matyax) [reply public/app/features/logs/components/PopoverMenu.tsx:62]: Will be addressed by https://github.com/grafana/grafana/pull/78322
